### PR TITLE
[sampling] feat: add Top-nσ logit truncation as built-in CustomLogitProcessor

### LINKED ISSUE #29119: [RFC] Add Top-nσ (Top-n-sigma) logit truncation as a built-in CustomLogitProcessor
## Motivation

[Top-nσ](https://arxiv.org/abs/2411.07641) ("Top-nσ: Not All Logits Are You Need", ACL 2025) is a logit-space dynamic truncation method that uses the **standard deviation** of the logit distribution to set a filtering threshold:

```
threshold = max_logit - n * std_logit
logits[logits < threshold] = -inf
```

Unlike probability-space filters (`top_p`, `min_p`), it operates **before softmax** (SGLang's custom logit processor pipeline already runs before temperature scaling), and adapts the candidate set to the "peakiness" of the distribution:

- **Sharp distribution** (model certain): small std → narrow threshold → fewer candidates
- **Flat distribution** (model uncertain): large std → wide threshold → more candidates

### How it compares to existing methods

| Method | Space | Truncation rule | Adaptive? |
|---|---|---|---|
| top_k | Probability | Keep top-k highest-probability tokens | ❌ Fixed count |
| top_p (nucleus) | Probability | Keep tokens until cumulative prob ≥ p | ✅ Adaptive count only |
| min_p | Probability | Keep tokens with prob ≥ max_prob × min_p | ✅ Relative threshold |
| top_nσ | **Logit** | threshold = max_logit − n × std | ✅ Dual-adaptive |

> Key distinction: top_k/top_p/min_p operate **after softmax** (probability space); top_nσ operates **before softmax** (logit space).

### Why Top-nσ?

- **Truncation before softmax** — top_nσ filters in logit space first, then softmax only processes the surviving candidates. In contrast, top_p/min_p must compute a full-vocab softmax first, then truncate.
- **Dual adaptivity** — the std-based threshold adjusts both *how many* and *how loosely* candidates are kept in one formula. top_k fixes the count; top_p adapts the count but not the threshold tightness.
- **Temperature-stable** — candidate count stays consistent across temperature scaling (logit std scales proportionally), while top_p/min_p candidate counts vary significantly with T.
- **Argmax-invariant** — the top-1 token always survives (0 std from itself), so greedy decoding is unaffected.

## Proposal

Add `TopNSigmaLogitProcessor` as a **built-in** `CustomLogitProcessor` in `python/sglang/srt/sampling/custom_logit_processor.py`, following the same pattern as the existing `ThinkingBudgetLogitProcessor` and `DisallowedTokensLogitsProcessor`.

### Implementation characteristics

- **Fully vectorized** across the batch (no per-row Python loop)
- **std computed once** and reused for both the `std==0` guard and the threshold formula (avoids a duplicate vocab-wide reduction)
- **Argmax-invariant** by construction

### Edge cases handled

| Case | Behavior |
|------|----------|
| `top_n_sigma <= 0` | Request skipped |
| `top_n_sigma` is None / non-numeric | Request skipped |
| `std == 0` (all logits equal) | Row skipped (no filtering) |
| NaN/Inf logits | Row skipped (finite guard) |
| Empty / None param list | Logits returned unchanged |

### Usage

```python
import requests
from sglang.srt.sampling.custom_logit_processor import TopNSigmaLogitProcessor

# Server must be started with --enable-custom-logit-processor
response = requests.post("http://localhost:30000/generate", json={
    "text": "The capital of France is",
    "sampling_params": {
        "temperature": 0.8,
        "custom_params": {"top_n_sigma": 2.0},
    },
    "custom_logit_processor": TopNSigmaLogitProcessor.to_str(),
})
```

## Scope

- New processor class in `custom_logit_processor.py` (~50 lines)
- Unit tests in `test/registered/unit/sampling/test_custom_logit_processor.py`
- No changes to `SamplingParams`, `SamplingBatchInfo`, or the sampler — uses the existing custom processor framework as-is

## Related PR

- #TBD — implementation

## References

Tang et al., "Top-nσ: Not All Logits Are You Need", ACL 2025. [arXiv:2411.07641](https://arxiv.org/abs/2411.07641)

### PR DESCRIPTION
**Pull request by Jackie2049**
## Summary

Closes #29119

Adds `TopNSigmaLogitProcessor` as a **built-in** `CustomLogitProcessor`, following the same pattern as the existing `ThinkingBudgetLogitProcessor` and `DisallowedTokensLogitsProcessor`.

Based on: Tang et al., "Top-nσ: Not All Logits Are You Need", ACL 2025. [arXiv:2411.07641](https://arxiv.org/abs/2411.07641)

### What is Top-nσ?

Top-nσ is a logit-space dynamic truncation method that uses the standard deviation of the logit distribution to set a filtering threshold:

```
threshold = max_logit - n * std_logit
logits[logits < threshold] = -inf
```

Unlike probability-space filters (`top_p`, `min_p`), it operates **before softmax** — which is exactly where SGLang's custom logit processor pipeline runs (`_preprocess_logits` is called before temperature scaling) — and adapts to distribution "peakiness":
- **Sharp distribution** (model certain): small std → narrow threshold → fewer candidates
- **Flat distribution** (model uncertain): large std → wide threshold → more candidates

### How it compares to other truncation methods

| Method | Space | Truncation rule | Adaptive? |
|---|---|---|---|
| top_k | Probability | Keep top-k highest-probability tokens | ❌ Fixed count |
| top_p (nucleus) | Probability | Keep tokens until cumulative prob ≥ p | ✅ Adaptive count only |
| min_p | Probability | Keep tokens with prob ≥ max_prob × min_p | ✅ Relative threshold |
| top_nσ | **Logit** | threshold = max_logit − n × std | ✅ Dual-adaptive |

> Key distinction: top_k/top_p/min_p operate **after softmax** (probability space); top_nσ operates **before softmax** (logit space).

### Why Top-nσ?

- **Truncation before softmax** — filters in logit space first, then softmax only processes the surviving candidates.
- **Dual adaptivity** — the std-based threshold adjusts both *how many* and *how loosely* candidates are kept in one formula.
- **Temperature-stable** — candidate count stays consistent across temperature scaling (logit std scales proportionally).
- **Argmax-invariant** — the top-1 token always survives, so greedy decoding is unaffected.

## Changes

- **New processor**: `TopNSigmaLogitProcessor` in `python/sglang/srt/sampling/custom_logit_processor.py`
- **Unit tests**: `TestTopNSigmaLogitProcessor` in `test/registered/unit/sampling/test_custom_logit_processor.py` (13 test cases)

### Implementation details

- **Fully vectorized** across the batch (no Python per-row loop)
- **std computed once** and reused for both the `std==0` guard and the threshold formula, avoiding a duplicate vocab-wide reduction
- **Argmax-invariant** by construction (the peak token always survives)

### Edge cases handled

| Case | Behavior |
|------|----------|
| `top_n_sigma <= 0` | Request skipped |
| `top_n_sigma` is None / non-numeric | Request skipped |
| `std == 0` (all logits equal) | Row skipped (no filtering) |
| NaN/Inf logits | Row skipped (finite guard) |
| Empty / None param list | Logits returned unchanged |

### Usage

Server must be started with `--enable-custom-logit-processor`:

```python
import requests
from sglang.srt.sampling.custom_logit_processor import TopNSigmaLogitProcessor

response = requests.post("http://localhost:30000/generate", json={
    "text": "The capital of France is",
    "sampling_params": {
        "temperature": 0.8,
        "custom_params": {"top_n_sigma": 2.0},
    },
    "custom_logit_processor": TopNSigmaLogitProcessor.to_str(),
})
```

No changes to `SamplingParams`, `SamplingBatchInfo`, or the sampler — uses the existing custom processor framework as-is.

## Testing plan

- [x] Unit tests cover: sharp/flat distributions, argmax invariance, zero-std, NaN/Inf, negative/None/zero n, mixed batch, empty params
- [ ] Manual end-to-end test with `--enable-custom-logit-processor` on a small model
- [ ] CI passes

Co-authored-by: Boundless <ruihang_wu@163.com>

---
### CI States

Latest PR Test (Base): :x: [Run #28076812108](https://github.com/sgl-project/sglang/actions/runs/28076812108)
Latest PR Test (Extra): :x: [Run #28076812084](https://github.com/sgl-project/sglang/actions/runs/28076812084)

### CODE REVIEW

**Comment by Jackie2049** [comment]
Closing — this was opened prematurely. Will reopen after review on the fork.
